bridge-ffi/src/lib.rs

```diff
@@ -0,0 +1,650 @@
+#![allow(non_camel_case_types)]
+
+use prost::Message;
+use std::str::FromStr;
+use temporal_sdk_core_protos::coresdk::bridge;
+
+extern crate libc;
+
+/// A set of bytes owned by Core. This must always be passed to
+/// tmprl_bytes_free when no longer in use.
+#[repr(C)]
+pub struct tmprl_bytes_t {
+    bytes: *const u8,
+    len: libc::size_t,
+    cap: libc::size_t,
+    disable_free: bool,
+}
+
+impl tmprl_bytes_t {
+    fn from_vec(vec: Vec<u8>) -> tmprl_bytes_t {
+        // Mimics Vec::into_raw_parts that's only available in nightly
+        let mut vec = std::mem::ManuallyDrop::new(vec);
+        tmprl_bytes_t {
+            bytes: vec.as_mut_ptr(),
+            len: vec.len(),
+            cap: vec.capacity(),
+            disable_free: false,
+        }
+    }
+
+    fn from_vec_disable_free(vec: Vec<u8>) -> tmprl_bytes_t {
+        let mut b = tmprl_bytes_t::from_vec(vec);
+        b.disable_free = true;
+        b
+    }
+
+    fn into_raw(self) -> *mut tmprl_bytes_t {
+        Box::into_raw(Box::new(self))
+    }
+}
+
+// We only impl this because they are required by lazy_static and raw pointers
+// are not usually safe for send
+unsafe impl Send for tmprl_bytes_t {}
+unsafe impl Sync for tmprl_bytes_t {}
+
+impl Drop for tmprl_bytes_t {
+    fn drop(&mut self) {
+        // In cases where freeing is disabled (or technically some other
+        // drop-but-not-freed situation though we don't expect any), the bytes
+        // remain non-null so we re-own them here
+        if !self.bytes.is_null() {
+            unsafe { Vec::from_raw_parts(self.bytes as *mut u8, self.len, self.cap) };
+        }
+    }
+}
+
+/// Free a set of bytes. The first parameter can be null in cases where a
+/// tmprl_core_t instance isn't available. If the second parameter is null, this
+/// is a no-op.
+#[no_mangle]
+pub extern "C" fn tmprl_bytes_free(core: *mut tmprl_core_t, bytes: *const tmprl_bytes_t) {
+    // Bail if freeing is disabled
+    unsafe {
+        if bytes.is_null() || (*bytes).disable_free {
+            return;
+        }
+    }
+    let bytes = bytes as *mut tmprl_bytes_t;
+    // Return vec back to core before dropping bytes
+    let vec = unsafe { Vec::from_raw_parts((*bytes).bytes as *mut u8, (*bytes).len, (*bytes).cap) };
+    // Set to null so the byte dropper doesn't try to free it
+    unsafe { (*bytes).bytes = std::ptr::null_mut() };
+    // Return only if core is non-null
+    if !core.is_null() {
```

Comment:
I'd maybe even panic if core is null here, it never should be

Reply:
Not true. For example, what if core init fails? The error is returned in a proto but core is nil, yet they should still free the proto. I even specifically call this out on. And really, I could remove this parameter altogether right now as we don't even have a byte pool (that's its only purpose).